src/extensions/default/PrefsCodeHints/main.js

```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright (c) 2015 Adobe Systems Incorporated. All rights reserved.
+ *  
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"), 
+ * to deal in the Software without restriction, including without limitation 
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, 
+ * and/or sell copies of the Software, and to permit persons to whom the 
+ * Software is furnished to do so, subject to the following conditions:
+ *  
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *  
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+ * DEALINGS IN THE SOFTWARE.
+ * 
+ */
+
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, indent: 4, maxerr: 50 */
+/*global define, brackets, $*/
+
+define(function (require, exports, module) {
+    "use strict";
+    
+    // Load dependencies.
+    var AppInit             = brackets.getModule("utils/AppInit"),
+        CodeHintManager     = brackets.getModule("editor/CodeHintManager"),
+        PreferencesManager  = brackets.getModule("preferences/PreferencesManager"),
+        StringMatch         = brackets.getModule("utils/StringMatch"),
+        ExtensionUtils      = brackets.getModule("utils/ExtensionUtils"),
+        EditorManager       = brackets.getModule("editor/EditorManager"),
+        LanguageManager     = brackets.getModule("language/LanguageManager"),
+        JSONUtils           = brackets.getModule("language/JSONUtils"),
+        Strings             = brackets.getModule("strings"),
+        data                = JSON.parse(require("text!data.json")),
+        isPrefDocument      = false,
+        isPrefHintsEnabled  = false;
+    
+    var stringMatcherOptions = {
+        preferPrefixMatches: true
+    };
+    
+    // List of parent keys for which no key hints will be provided.
+    var parentKeyBlacklist = [
+        "language.fileExtensions",
+        "language.fileNames",
+        "path"
+    ];
+    
+    // Define a preference for code hinting.
+    PreferencesManager.definePreference("codehint.PrefHints", "boolean", true);
```

Comment:
How about more argument is added to the definePreference API and (maybe the last parameter as description/ object with possible values for the prefs) and then raise a `preferences- code hints missing` warning so that extension devs can directly incorporate code hints to their preferences?